lib/cretonne/src/isa/intel/mod.rs

```diff
@@ -111,4 +117,149 @@ impl TargetIsa for Isa {
+            ));
+        }
+
+        for csr_arg in &csr_args {
+            func.signature.params.push(*csr_arg);
+        }
+        for csr_arg in csr_args.iter().rev() {
+            func.signature.returns.push(*csr_arg);
+        }
+
+        // Append param to entry EBB
+        let entry_ebb = func.layout.entry_block().expect("missing entry block");
+        func.dfg.append_ebb_param(entry_ebb, csr_type);
+
+        // Find our frame pointer parameter Value
+        let fp = func.special_param(ir::ArgumentPurpose::FramePointer)
+            .expect("missing frame pointer");
+
+        // Assign it a location
+        func.locations[fp] = ir::ValueLoc::Reg(RU::rbp as RegUnit);
+
+        let mut csr_vals = Vec::new();
+        for reg in &csrs {
+            // Append param to entry EBB
+            func.dfg.append_ebb_param(entry_ebb, csr_type);
+
+            let csr_arg = func.dfg.ebb_params(entry_ebb).last().expect(
+                "no last argument",
+            );
+
+            // Assign it a location
+            func.locations[*csr_arg] = ir::ValueLoc::Reg(*reg as RegUnit);
+
+            // Remember it so we can push it momentarily
+            csr_vals.push(*csr_arg);
+        }
+
+
+        // Insert prologue
+        {
+            let mut pos = EncCursor::new(func, self).at_first_insertion_point(entry_ebb);
+            pos.ins().x86_push(fp);
+            pos.ins().copy_special(
+                RU::rsp as RegUnit,
+                RU::rbp as RegUnit,
+            );
+            if local_stack_size > 0 {
+                pos.ins().adjust_sp_imm(-(local_stack_size as i32));
+            }
+
+            for csr_arg in csr_vals {
+                pos.ins().x86_push(csr_arg);
+            }
+        }
+
+        // Find all 'return' instructions
+        let mut return_insts = Vec::new();
+        for ebb in func.layout.ebbs() {
+            for inst in func.layout.ebb_insts(ebb) {
+                if let InstructionData::MultiAry { opcode, .. } = func.dfg[inst] {
+                    if opcode == Opcode::Return {
+                        return_insts.push(inst);
+                    }
+                }
+
+            }
+        }
+
+        // Insert an epilogue directly before every 'return'
+        for inst in return_insts {
+            self.insert_epilogue(inst, local_stack_size as i32, func, &csrs, csr_type);
+        }
+
+
+        Ok(())
+    }
+}
+
+impl Isa {
+    fn insert_epilogue(
+        &self,
+        inst: ir::Inst,
+        stack_size: i32,
+        func: &mut ir::Function,
+        csrs: &Vec<RU>,
+        csr_type: ir::types::Type,
+    ) {
+        let mut return_values = Vec::new();
+
+        let mut pos = EncCursor::new(func, self).at_inst(inst);
+        if stack_size > 0 {
+            pos.ins().adjust_sp_imm(stack_size);
+        }
+        for reg in csrs.iter().rev() {
+            let csr_ret = pos.ins().x86_pop(csr_type);
+            return_values.push((csr_ret, *reg));
```

Comment:
If you need to, you can insert instructions backwards by using `pos.prev_inst()` after inserting each pop. This might come up if you are trying to get rid of vectors in these algorithms.